tracer/src/Datadog.Trace.SourceGenerators/PublicApi/PublicApiGenerator.cs

```diff
@@ -0,0 +1,273 @@
+            var source = Sources.CreatePartialClass(sb, nameSpace, className, partialClass);
+            context.AddSource($"{nameSpace}.{className}.g.cs", SourceText.From(source, Encoding.UTF8));
+        }
+    }
+
+    private static Result<(PublicApiProperty PropertyTag, bool IsValid)> GetPublicApiProperties(
+        GeneratorAttributeSyntaxContext ctx, CancellationToken ct)
+    {
+        var field = (FieldDeclarationSyntax)ctx.TargetNode.Parent!.Parent!;
+        var classDec = field.Parent as ClassDeclarationSyntax;
+        if (classDec is null)
+        {
+            // only support fields on classes
+            return new Result<(PublicApiProperty PropertyTag, bool IsValid)>(
+                (default, false),
+                new EquatableArray<DiagnosticInfo>(new[] { OnlySupportsClassesDiagnostic.CreateInfo(field) }));
+        }
+
+        var fieldSymbol = ctx.TargetSymbol as IFieldSymbol;
+        if (fieldSymbol is null)
+        {
+            // something weird going on
+            return new Result<(PublicApiProperty PropertyTag, bool IsValid)>((default, false), default);
+        }
+
+        List<DiagnosticInfo>? diagnostics = null;
+        bool hasMisconfiguredInput = false;
+        int? publicApiGetter = null;
+        int? publicApiSetter = null;
+
+        foreach (AttributeData attributeData in fieldSymbol.GetAttributes())
+        {
+            if ((attributeData.AttributeClass?.Name == "GeneratePublicApiAttribute" ||
+                 attributeData.AttributeClass?.Name == "GeneratePublicApi")
+             && attributeData.AttributeClass.ToDisplayString() == GeneratePublicApiAttribute)
+            {
+                var args = attributeData.ConstructorArguments;
+                if (args.Length == 0)
+                {
+                    hasMisconfiguredInput = true;
+                    break;
+                }
+
+                foreach (TypedConstant typedConstant in attributeData.ConstructorArguments)
+                {
+                    if (typedConstant.Kind == TypedConstantKind.Error)
+                    {
+                        hasMisconfiguredInput = true;
+                        break;
+                    }
+                }
+
+                if (hasMisconfiguredInput)
+                {
+                    break;
+                }
+
+                publicApiGetter = args[0].Value as int?;
+
+                if (args.Length > 1)
+                {
+                    publicApiSetter = args[1].Value as int?;
+                    if ((fieldSymbol.IsReadOnly || fieldSymbol.IsConst) && publicApiSetter.HasValue)
+                    {
+                        diagnostics ??= new List<DiagnosticInfo>();
+                        diagnostics.Add(SetterOnReadonlyFieldDiagnostic.CreateInfo(attributeData.ApplicationSyntaxReference?.GetSyntax()));
+                        hasMisconfiguredInput = true;
+                    }
+                }
+
+                break;
```

Comment:
Yeah, once we find the `[GeneratePublicApi]` attribute we're "done", so there's no point checking any other attributes it might have. It's a tiny optimisation

Reply:
Ahhhh okay that makes sense thanks